miner: Add block building interruption on payload resolution (getPayload)

sebastianst opened the pull request:
**Description**

- Block building is interrupted on a `getPayload` engine api call.
- Empty block is now only built for `NoTxPool` FCU calls, and as before, FCU blocks on building the empty block. This has the advantage that if generating a lot of empty blocks, `getPayload` can immediately be called after `forkchoiceUpdated` to return that empty block.
- If using the tx pool, FCU doesn't block on any block building but returns immediately after creating the payload promise.
    - Resolution of the payload promise (`getPayload`) now waits for the first full block to be built, as there is no empty block to return. If at least one block has already been built, as before, it is returned immediately, while block building is interrupted in the background.

**Tests**

Extended the `TestBuildPayload` test to cover both cases of `NoTxPool`.

Added an ugly global bool to indicate that a test is running, which guarantees one full block is built and block building is not interrupted. This seemed to be the least-invasive way to make all tests still pass. This was necessary because a couple of test call `Resolve` or `getPayload` immediately after a FCU, but this would now interrupts the block building process, which made tests fail.

**Additional context**



**Metadata**

- Fixes https://github.com/ethereum-optimism/client-pod/issues/173
- Fixes https://github.com/ethereum-optimism/client-pod/issues/187
- Fixes https://github.com/ethereum-optimism/client-pod/issues/163

sebastianst commented on miner/payload_building.go:
```diff
@@ -191,12 +213,14 @@ func (payload *Payload) ResolveFull() *engine.ExecutionPayloadEnvelope {
+	payload.interrupt.Store(commitInterruptResolve)
+}
+
+func (w *worker) emptyBlock(args *BuildPayloadArgs) *newPayloadResult {
```
I might put this back into `buildPayload` if we don't need it in the payload updating goroutine. Initially moved it here into its own function to avoid code duplication.

sebastianst commented on miner/payload_building.go:
```diff
@@ -209,23 +233,62 @@ func (w *worker) buildPayload(args *BuildPayloadArgs) (*Payload, error) {
 		return payload, nil
 	}
 
+	fullParams := &generateParams{
```
It's not ideal to pull the definition of `fullParams` out of the goroutine, because it increases the diff.

But we would miss parameter validation otherwise, so need to do this.

I caught this because the test `TestInvalidPayloadTimestamp` was failing.

sebastianst commented on miner/worker.go:
```diff
@@ -937,8 +939,34 @@ type generateParams struct {
+// validateParams validates the given parameters.
+// It currently checks that the parent block is known and that the timestamp is valid,
+// i.e., after the parent block's timestamp.
+func (w *worker) validateParams(genParams *generateParams) error {
```
This is largely copied from the beginning of `prepareWork`.

An alternative would be to extract this part out of `prepareWork` and reuse it there, but then we would also need to return the `parent` and `timestamp`, which I found cumbersome, so opted for the copy.

sebastianst reviewed (commented):
noted from call with proto

sebastianst commented on miner/payload_building.go:
```diff
@@ -235,19 +298,7 @@ func (w *worker) buildPayload(args *BuildPayloadArgs) (*Payload, error) {
+		defer endTimer.Stop()
 
 		for {
 			select {
```
or log in `payload.update`

sebastianst replied:
added error logging to `payload.update`.

sebastianst commented on miner/payload_building.go:
```diff
@@ -142,9 +147,19 @@ func (payload *Payload) update(r *newPayloadResult, elapsed time.Duration) {
 
+	// Interrupt payload generation
+	payload.interruptWorker()
+	// TODO: Since we don't wait for the next payload update via payload.cond.Wait(),
```
done, added `payload.done` channel which the routine closes on return.

sebastianst commented on miner/payload_building.go:
```diff
@@ -91,6 +92,8 @@ type Payload struct {
 	stop     chan struct{}
 	lock     sync.Mutex
 	cond     *sync.Cond
+
```
logging errors for now and just returning nil, as before

sebastianst commented on miner/payload_building.go:
```diff
@@ -182,47 +193,87 @@ func (payload *Payload) ResolveFull() *engine.ExecutionPayloadEnvelope {
+	} else if !onlyFull && payload.empty != nil {
+		return engine.BlockToExecutableData(payload.empty, big.NewInt(0), nil)
+	} else if err := payload.err; err != nil {
+		log.Error("Error building any payload", "id", payload.id, "err", err)
```
The latest payload building error is now at least bubbled up to here, if no block could be built at all.

ajsutton reviewed (commented):
When talking about returning the "full" block, are we literally waiting for it to be packed with as many tx as possible, or does the interrupt stop adding tx to the block?

The ideal behaviour when `getPayload` is called is basically:
* If a valid block has already been built and is ready, return that and discard the in-progress block.
* Otherwise, stop adding tx from the pool immediately and finalise the block in progress so it can be returned.

Potentially you could be even smarter and compare the value of the existing block to the value of the block in progress and finalized the block in progress if it's already worth more.

That may mean that we return a half-full block because that's all we had time to build but that's a lot better than being late delivering the block. I wonder if that potentially simplifies some of the logic here as well since FCU would always defer building to the worker thread and `getPayload` would always have a block to return.

sebastianst commented:
> The ideal behaviour when `getPayload` is called is basically:
> 
> * If a valid block has already been built and is ready, return that and discard the in-progress block.
> * Otherwise, stop adding tx from the pool immediately and finalise the block in progress so it can be returned.

@ajsutton It is implemented like this 👍🏻 

I now realize that there's one thing that can be optimized: in `worker.generateWork`, return immediately after `fillTransactions` if an interrupt is set _and_ one block is already available. The current implementation still calls `FinalizeAndAssemble` even if the result is gonna be discarded. Will add this.

> Potentially you could be even smarter and compare the value of the existing block to the value of the block in progress and finalized the block in progress if it's already worth more.

> That may mean that we return a half-full block because that's all we had time to build but that's a lot better than being late delivering the block. I wonder if that potentially simplifies some of the logic here as well since FCU would always defer building to the worker thread and getPayload would always have a block to return.

This is exactly the implementation that I had up until last week. We then discussed last week that we rather want `getPayload` to instantly return if at least one block is already done, instead of it blocking on the currently ongoing block building process. This convinced me because my investigations showed that applying a single transaction, or calling `FinalizeAndAssemble`, might itself take several seconds in some instances.

The process could also happen with a timeout. Say, wait up to 10ms for the interrupted block building process to complete and the comparison to be done, and then return the existing block already if this times out. However, I remember that we discussed that an instant `getPayload` return is preferred.

If we want to revisit that decision, I suggest to do that in a follow-up.

protolambda commented on eth/catalyst/api_test.go:
```diff
@@ -51,6 +51,10 @@ import (
 )
 
+func init() {
+	miner.IsPayloadBuildingTest = true
```
I'm not a fan of the global parameter either. Can we not pass it through a `context` or something, as an optional context value?

ajsutton replied:
Given https://github.com/ethereum-optimism/op-geth/blob/c51655402f900cdafdd86f9dab84a13f735f36ca/eth/catalyst/api_test.go#L206 I wonder if these tests were always flaky - a fixed time based wait is never a good sign.

Yep - flaky even without these changes as they may return an empty block and fail. And they use different amounts of time to wait in different tests... 

I had a play with it in https://github.com/ethereum-optimism/op-geth/pull/199 - it's still an arbitrary sleep but it's 1 second.  There's only a couple of tests that need it and so it's better to have a long period and less flakiness.  The tests do have an ethservice which can get to the miner so it may be possible to add an API to check if the full block has been built yet and wait deterministically but I didn't try. That's passed in CI and passed consistently locally.

sebastianst replied:
Thanks, merged #199 into this PR.

@protolambda it wasn't possible to pass anything down in tests in general because some tests would just call fcu on the engine API or similar, which don't have a context.

sebastianst replied:
Merged in #204 to address this.

ajsutton reviewed (commented):
Ok, went through this carefully and it looks good to me. Only question is really around the global var and as mentioned, I think we can avoid it but you may have seen something I missed.

sebastianst commented:
@ajsutton thanks for review, I'm happy with your test workaround as well, will merge it into this pr

sebastianst commented on miner/payload_building.go:
```diff
@@ -91,6 +95,10 @@ type Payload struct {
+
+	err       error
+	stopOnce  sync.Once
+	interrupt *atomic.Int32 // interrup signal shared with worker
```
fixed by merging #199

protolambda commented:
Reviewed the payload-interrupt code, and the discussion about the half-full block, but I think we are not interrupting correctly on the first payload currently:

- If `resolve(false)`, and if there's no block yet, it currently does not interrupt the block building at all: it waits for the payload with `payload.cond.Wait()`, and only then calls `payload.stopBuilding()`.
- If an interrupt is received by the worker, during `fillTransactions`, then it doesn't seal the work, and returns an error instead. The error is propagated to `payload.update()`, and would set the `payload.cond`, but due to order of operations it'll be waiting for this `cond` before firing the interrupt. 

The above two things combined mean we can't fire `stopBuilding()` early (due to lack of sealing on interrupt error)), and are forcing ourselves to build the full 30M gas block (due to lack of interrupt, since we wait for the `cond` if it's the first block).

I think we should add a flag to `generateParams` that defines whether or not we want to seal the half-full block after interrupt, or return an interrupt error.
And then run `stopBuilding()` before running the `payload.cond.Wait()`, to get that half-full block result, rather than not interrupting the first block. (This should work fine, since the interrupt is clean (we stop adding more txs, it doesn't leave the block-building state in an inconsistent place).

That way we don't get hung up on large 30M block building attempts if it's the first block building attempt.

sebastianst commented:
Thanks Proto, I moved the interrupt now before potentially waiting on the conditional, so we interrupt even the first block building routine. It was kind of a deliberate decision at the time to build the first block fully, but it's indeed better to even interrupt the first block, because even that one may take a long time.

As you wrote in the edit, I already added the `isUpdate` flag to `generateParams` which signals whether we seal an interrupted block or not.

I added another test that confirms that interruption actually works.

I now had to add more sleeps to the `eth/catalyst/api_test.go` tests, which now lets the tests run for ~**143 s** (measured with `time go test ./eth/catalyst -count 1`) instead of **13**s! This was the reason in the first place why I just disabled interrupts completely in tests with the global var hack, because all geth tests assume that resolve doesn't interrupt. I personally would still favor just disabling it for the tests in `eth/catalyst` rather than adding sleeps.

ajsutton commented:
hmm, so there's a failure in `TestSimulatedBeaconSendWithdrawals` which is using `GetPayload` straight after starting block building (https://github.com/ethereum-optimism/op-geth/blob/a8a9c8e4b00c5b9f84242181839234b8e9fd54e3/eth/catalyst/simulated_beacon.go#L153-L172) so that's another place we'd need to wait for the building to complete.

I've exposed a couple of methods to deterministically wait for the block building to complete rather than needing a fixed timeout in https://github.com/ethereum-optimism/op-geth/pull/204 and hooked that up in the simulated beacon as well.  We may need to add appropriate calls to wait in more tests yet which will be an annoying source of flakiness for a while but I think this is the right approach to deterministically wait.  Definitely open to better ways to expose the ability to wait though.

sebastianst commented:
Refined @ajsutton's solution and fixed a buggy test (`TestNilWithdrawals`)

protolambda commented on miner/worker.go:
```diff
@@ -1095,8 +1128,15 @@ func (w *worker) generateWork(genParams *generateParams) *newPayloadResult {
 		if errors.Is(err, errBlockInterruptedByTimeout) {
 			log.Warn("Block building is interrupted", "allowance", common.PrettyDuration(w.newpayloadTimeout))
+		} else if errors.Is(err, errBlockInterruptedByResolve) {
+			timer.Stop() // don't need timeout interruption any more
```
Maybe it's better to just always call `timer.Stop()` after `fillTransactions`, to cover the success-case before timeout also? And then remove the `defer timer.Stop()`

sebastianst replied:
Done. Wanted to keep the diff minimal, but that also works and is cleaner.

protolambda commented on miner/payload_building.go:
```diff
@@ -231,38 +313,61 @@ func (w *worker) buildPayload(args *BuildPayloadArgs) (*Payload, error) {
 		defer timer.Stop()
 
+		start := time.Now()
+		const blockTime = 2 * time.Second // OP default block time
 		// Setup the timer for terminating the process if SECONDS_PER_SLOT (12s in
 		// the Mainnet configuration) have passed since the point in time identified
 		// by the timestamp parameter.
-		endTimer := time.NewTimer(time.Second * 12)
+		endTimer := time.NewTimer(blockTime)
+		defer endTimer.Stop()
```
This code-path shouldn't be dependent on a constant block-time, especially since it's also used for L1 testing in op-e2e and possibly elsewhere.

Instead, can we make `validateParams` return the block time? It already looks up the parent block header (which includes the timestamp), and has the payload arguments with timestamp, so it can compute the difference. And then maybe clip it to a minimum of `2s` (in case the block-time is unreasonably low, or instant).

protolambda reviewed (commented):
Looks good! Sorry I missed a few minor things in my last review, the PR looks ready otherwise
